### PR TITLE
This tightens the AST compact printing

### PR DESCRIPTION
**Pull request by bbakerman**
See #2869

Tests need to be fixed

### CODE REVIEW

File: src/main/java/graphql/language/AstPrinter.java [hunk 1]
```diff
@@ -523,7 +547,35 @@ private <T extends Node> String join(List<T> nodes, String delim) {
         return join(nodes, delim, "", "");
     }
 
+    /*
+     * Some joined nodes don't need delimiters between them and some do
+     * This encodes that knowledge of those that don't require delimiters
+     */
     @SuppressWarnings("SameParameterValue")
+    private <T extends Node> String jointight(List<T> nodes, String delim, String prefix, String suffix) {
```
Comment (dondonz) [inline src/main/java/graphql/language/AstPrinter.java:555]: Hyper nit: could you make it camelCase `joinTight`

For people with poor eyesight like me

File: src/main/java/graphql/language/AstPrinter.java [hunk 2]
```diff
@@ -523,7 +547,35 @@ private <T extends Node> String join(List<T> nodes, String delim) {
...
+        StringBuilder joined = new StringBuilder();
+        joined.append(prefix);
+
+        String lastNodeText = "";
+        boolean first = true;
+        for (T node : nodes) {
+            if (first) {
+                first = false;
+            } else {
+                boolean canButtTogether = lastNodeText.endsWith("}");
```
Comment (dondonz) [inline src/main/java/graphql/language/AstPrinter.java:565]: Did you mean `canPutTogether`?

**Review by dondonz — APPROVED** [review]

File: src/test/groovy/graphql/language/AstPrinterTest.groovy [hunk 3]
```diff
@@ -583,7 +625,17 @@ extend input Input @directive {
         String output = AstPrinter.printAstCompact(document)
 
         expect:
-        output == '''{aliasOfFoo:foo(arg1:"val1",args2:"val2") @isCached {hello} world @neverCache @okThenCache} fragment FX on SomeType {aliased:field(withArgs:"argVal",andMoreArgs:"andMoreVals")}'''
+        isParseableAst(output)
+        output == '''{aliasOfFoo:foo(arg1:"val1",args2:"val2") @isCached{hello}world @neverCache @okThenCache} fragment FX on SomeType {aliased:field(withArgs:"argVal",andMoreArgs:"andMoreVals")}'''
+    }
```
Comment (tomgasson) [inline src/test/groovy/graphql/language/AstPrinterTest.groovy:630]: You can also tighten directives

`{foo(arg1:"val1",args2:"val2") @isCached{hello}world @neverCache @okThenCache}}`

`{foo(arg1:"val1",args2:"val2")@isCached{hello}world@neverCache@okThenCache}}`

File: src/main/java/graphql/language/AstPrinter.java (hunk 2 above)
Comment (dondonz) [reply src/main/java/graphql/language/AstPrinter.java:565]: Ok you win this time

**Pull request merged**